Products/CMFCore/indexing.py

```diff
@@ -0,0 +1,330 @@
+        # TODO: must the queue be handled independently for each processor?
+        for op, obj, attributes, metadata in self.queue:
+            for name, util in utilities:
+                if op == INDEX:
+                    util.index(obj, attributes)
+                elif op == REINDEX:
+                    util.reindex(obj, attributes, update_metadata=metadata)
+                elif op == UNINDEX:
+                    util.unindex(obj)
+                else:
+                    raise InvalidQueueOperation(op)
+            processed += 1
+        debug('finished processing %d items...', processed)
+        self.clear()
+        return processed
+
+    def commit(self):
+        sm = getSiteManager()
+        for name, util in sm.getUtilitiesFor(IIndexQueueProcessor):
+            util.commit()
+
+    def abort(self):
+        sm = getSiteManager()
+        for name, util in sm.getUtilitiesFor(IIndexQueueProcessor):
+            util.abort()
+        self.clear()
+
+    def clear(self):
+        del self.queue[:]
+        # release transaction manager
+        self.tmhook = None
+
+
+def filterTemporaryItems(obj, checkId=True):
+    """ check if the item has an acquisition chain set up and is not of
+        temporary nature, i.e. still handled by the `portal_factory`;  if
+        so return it, else return None """
+    parent = aq_parent(aq_inner(obj))
+    if parent is None:
+        return None
+    if IBrowserRequest.providedBy(parent):
+        return None
+    if checkId and getattr(obj, 'getId', None):
+        parent = aq_base(parent)
+        if getattr(parent, '__contains__', None) is None:
+            return None
+        elif obj.getId() not in parent:
+            return None
+    isTemporary = getattr(obj, 'isTemporary', None)
+    if isTemporary is not None:
+        try:
+            if obj.isTemporary():
+                return None
+        except TypeError:
+            return None # `isTemporary` on the `FactoryTool` expects 2 args
+    return obj
+
+
+class QueueSavepoint(object):
+    """ transaction savepoints using the IIndexQueue interface """
+
+    def __init__(self, queue):
+        self.queue = queue
+        self.state = queue.getState()
+
+    def rollback(self):
+        self.queue.setState(self.state)
+
+
+@implementer(ISavepointDataManager)
+class QueueTM(local):
+    """ transaction manager hook for the indexing queue """
+
+    def __init__(self, queue):
+        local.__init__(self)
+        self.registered = False
+        self.vote = False
+        self.queue = queue
+
+    def register(self):
+        if not self.registered:
+            try:
+                transaction = getTransaction()
+                transaction.join(self)
+                transaction.addBeforeCommitHook(self.before_commit)
+                self.registered = True
+            except Exception:
+                logger.exception('Exception during register (registered=%s)',
+                    self.registered)
```

Comment:
Swallowing exceptions raised by the transaction doesn't seem like the Right Thing  here.